File: Day66/Part1/mana_drain.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

bool solve() {
    int n, M;
    if (!(cin >> n >> M)) return false;
    vector<int> drain(n);
    for (int i = 0; i < n; ++i) cin >> drain[i];

    set<int> visited;
    int current_pos = 0;

    while (M > 0) {
        if (visited.count(current_pos)) return true;
        visited.insert(current_pos);

        M -= drain[current_pos];
        if (M <= 0) return false;

        current_pos = (current_pos + drain[current_pos]) % n;
    }
    return false;
}

int main() {
    if (solve()) cout << "True" << endl;
    else cout << "False" << endl;
    return 0;
}

File: Day66/Part1/mana_drain.py
def solve():
    try:
        line1 = input().split()
        if not line1: return
        n = int(line1[0])
        M = int(line1[1])
        
        drain = list(map(int, input().split()))
        
        visited = set()
        current_pos = 0
        
        while M > 0:
            if current_pos in visited:
                print("True")
                return
            visited.add(current_pos)
            
            M -= drain[current_pos]
            if M <= 0:
                print("False")
                return
            
            current_pos = (current_pos + drain[current_pos]) % n
            
        print("False")
    except EOFError:
        pass

if __name__ == "__main__":
    solve()
